karapace/kafka_rest_apis/__init__.py

```diff
@@ -40,36 +49,28 @@ def __init__(self, config: Config) -> None:
         super().__init__(config=config)
         self._add_kafka_rest_routes()
         self.serializer = SchemaRegistrySerializer(config=config)
-        self._cluster_metadata = None
-        self._metadata_birth = None
-        self.metadata_max_age = self.config["admin_metadata_max_age"]
-        self.admin_client = None
-        self.admin_lock = None
-        self.metadata_cache = None
-        self.schemas_cache = {}
-        self.consumer_manager = ConsumerManager(config=config)
-        self.init_admin_client()
+        self.proxies: Dict[str, "UserRestProxy"] = {}
```

Comment:
Any consideration to have cleanup on rarely used proxies?

Reply:
And maybe set a high enough max number on how many proxies there can be to prevent excessive resource usage? And when that is reached and a new one is created, close and remove one of the proxies, maybe in the oldest in LRU order?

Reply:
Only one proxy per **valid** user, password pair is ever created. (Thus password change for user will create a new one.) Therefore I don't know if a max limit is really needed as number of users in Kafka already is one limit. And in any case removal can be done only if there is no active consumers. But of course passive proxies keep Kafka connection open, thus disconnecting those (if only producer/metadata fetch, not consumer) does release some resources.

Reply:
What about caching per `(user,password)`. And when a new entry is created for a `user` (with a valid user + password tuple), the older ones are invalid because they use an old password. And since they're invalid, could they be purged?

Reply:
There is now janitor task for idle proxies. It will release proxy after five minutes of idle if the proxy does not have consumers. If the proxy has consumers, then there is a configuration value which destructs consumers, the default for this is one hour, but feature can be disabled with the configuration value.